src/v/kafka/client/direct_consumer/fetcher.cc

```diff
@@ -0,0 +1,538 @@
+}
+
+ss::future<> fetcher::do_fetch() {
+    bool needs_backoff = false;
+    try {
+        co_await _partitions_updated.wait(
+          [this] { return !_partitions.empty(); });
+
+        // this is the place where the loop will be stopped when the parent
+        // consumer updates fetchers state.
+        auto lock = co_await lock_state();
+        /**
+         * Iterate once over all partitions that are assigned to the fetcher and
+         * collect necessary actions. f.e. list offsets or include in fetch
+         * request.
+         */
+        auto partitions = co_await collect_partitions();
+
+        auto list_offsets_err = co_await maybe_initialise_fetch_offsets(
+          partitions);
+        /**
+         */
+        if (list_offsets_err != kafka::error_code::none) {
+            vlog(
+              logger().debug,
+              "[broker: {}] list offsets error: {}",
+              _id,
+              list_offsets_err);
+            if (is_retriable_error(list_offsets_err)) {
+                needs_backoff = true;
+            } else {
+                // propagate not retriable error to the queue
+                co_await queue().push_error(list_offsets_err);
+            }
+        }
+
+        auto req = co_await make_fetch_request(std::move(partitions));
+
+        if (_as.abort_requested()) {
+            // if the abort was requested, we should not dispatch the request
+            // and just return
+            co_return;
+        }
+        // TODO: cache the version of the fetch request
+        auto version = co_await get_fetch_request_version();
+        auto response = co_await _parent->_cluster->dispatch_to(
+          _id, std::move(req), version);
+        auto fetch_result = co_await process_fetch_response(
+          std::move(response));
+
+        if (fetch_result.has_error()) {
+            if (is_retriable_error(fetch_result.error())) {
+                needs_backoff = true;
+            } else {
+                // propagate non retriable error to the queue
+                auto f = queue().push_error(fetch_result.error());
+                lock.return_all();
+                co_await std::move(f);
+                co_return;
+            }
+        }
+
+        auto fetch_result_value = std::move(fetch_result.value());
+        if (fetch_result_value.needs_metadata_update) {
+            // if we need to update metadata, we should do it
+            // so that we can retry fetching the partitions later
+            needs_backoff = true;
+        }
+
+        if (!fetch_result_value.topics.empty()) {
+            auto f = queue().push(
+              std::move(fetch_result_value.topics),
+              fetch_result_value.total_bytes);
+            // we do not need to hold the lock after we have pushed the data
+            // to the queue, data order is now locked, we can release the lock
+            // and wait for the space in queue
+            lock.return_all();
```

Comment:
what exactly is the lock guaranteeing?

Reply:
There are two locks that are being acquired by the fetcher. One is required for correctness the other one i added to make the concurrency model safer. 
First is the shared read-write lock that is owned by the `direct_consumer` this one is being used to synchronize all fetchers and introduce a kind of stop-the-world barrier when fetcher assigned partitions must be updated. This way we guarantee that partition is only assigned to one fetcher and the fetcher is done fetching the partition (there is no inflight request in any of the fetchers) (we can relax this in future by using multiple locks if we need to). The other lock in fetcher is there just to make sure the partition state is not modified while fetching so it is easier to reason about the concurrency.